Document query scheduler

pstibrany opened the pull request:
**What this PR does**: This PR adds section about query scheduler to architecture overview.

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,15 +5,35 @@ weight: 5
+
+## Using Query Scheduler
+
+Query scheduler is new service that moves the in-memory queue from query frontend to a separate component.
```
Remember doc is here to stay. Try to avoid time-based references, like "new" or "now".

```suggestion
Query scheduler is a service that moves the in-memory queue from query frontend to a separate component.
```

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,15 +5,35 @@ weight: 5
+
+Query scheduler is new service that moves the in-memory queue from query frontend to a separate component.
+This makes scaling query frontend easier, as it allows running multiple query frontends without increasing the number of queues,
+which makes life difficult for queriers (as above-linked document explains).
```
> as above-linked document explains

I would just re-link it. BTW, links in the doc should be relative and not absolute.

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,15 +5,35 @@ weight: 5
+When using query frontend (possibly in combination with query scheduler), queriers need to connect to frontends or schedulers.
+To make sure that querier doesn't receive more queries that it can handle at the same time, make sure to configure the querier to match its PromQL concurrency with number of connections.
+This allows the operator to freely scale the frontend or scheduler up and down without impacting the amount of work an individual querier is attempting to perform.
+More details [here](https://cortexmetrics.io/docs/proposals/scalable-query-frontend/#dynamic-querier-concurrency).
```
I'm a bit unsure about linking proposals in the doc. The reason is that we don't keep proposals updated (and we shouldn't) but proposals get outdated over the time. Linking to proposals may be misleasing.

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,15 +5,35 @@ weight: 5
 
-Make sure to configure the querier frontend worker to match max concurrency.  This will allow the operator to freely scale the frontend up and down without impacting the amount of work an individual querier is attempting to perform.  More details [here](https://cortexmetrics.io/docs/proposals/scalable-query-frontend/#dynamic-querier-concurrency).
+When using query frontend (possibly in combination with query scheduler), queriers need to connect to frontends or schedulers.
+To make sure that querier doesn't receive more queries that it can handle at the same time, make sure to configure the querier to match its PromQL concurrency with number of connections.
```
> make sure to configure the querier to match its PromQL concurrency with number of connections

Could you mention the involved CLI flags? May not be clear to the reader.

pstibrany replied:
There is an example after this paragraph.

pracucci reviewed (approved):
Thanks @pstibrany for working on the doc! LGTM. I just left minor comments.

pstibrany commented:
Thanks for your review @pracucci. I've now reworked "Scaling the Query Frontend":
- added section about difficulties of scaling query frontend (copied and modified from proposal)
- inlined config options
- expanded Query scheduler section
- removed links to proposal, and only left one on the top (for "historical" context)
- made links relative

PTAL when you find some time. Thank you!

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,25 +5,52 @@ weight: 5
+## Scaling the Query Frontend
 
-When a new frontend is first created on scale up it will not immediately have queriers attached to it.  The existing endpoint `/ready` was updated to only return http 200 when the query frontend was ready to serve queries.  Make sure to configure this endpoint as a healthcheck in your load balancer.  Otherwise a query frontend scale up event might result in failed queries or high latency for a bi ...
+For every query frontend the querier adds a [configurable number of goroutines](https://github.com/cortexproject/cortex/blob/1797adfed2979f6096c3305b0dc9162c1ec0c046/pkg/querier/worker/worker.go#L212)
```
You know what I think about mentioning goroutines in the doc :) I don't think our doc should either talk about goroutines or link to code.

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,25 +5,52 @@ weight: 5
+(using `-frontend.scheduler-address` and `-querier.scheduler-address` options respectively).
+
+Note that querier will only fetch queries from query frontend OR query scheduler, but not both.
+If configured with both `-querier.frontend-address` and `-querier.scheduler-address`, querier will only use scheduler address.
```
We should add a config validation to make it fail if both addresses are specified. What do you think?

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,25 +5,52 @@ weight: 5
+If configured with both `-querier.frontend-address` and `-querier.scheduler-address`, querier will only use scheduler address.
+
+When using query scheduler, it is recommended to run two query scheduler instances.
+Running only one query scheduler poses a risk of increased query latency during scheduler restarts.
```
Also for HA purposes.

pstibrany replied:
> ... when single scheduler crashes or restarts.

OK?

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,25 +5,52 @@ weight: 5
+### DNS Configuration / Readiness
+
+When a new frontend is first created on scale up it will not immediately have queriers attached to it.
+The existing endpoint `/ready` was updated to only return http 200 when the query frontend was ready to serve queries.
```
I would not mention "was updated" but I would say:

```suggestion
The query-frontend `/ready` endpoint returns HTTP 200 status code only when the query-frontend is ready to serve queries.
```

pracucci reviewed (approved):
Good job! Way better than before. I left few more comments.

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,25 +5,52 @@ weight: 5
-When a new frontend is first created on scale up it will not immediately have queriers attached to it.  The existing endpoint `/ready` was updated to only return http 200 when the query frontend was ready to serve queries.  Make sure to configure this endpoint as a healthcheck in your load balancer.  Otherwise a query frontend scale up event might result in failed queries or high latency for a bi ...
+For every query frontend the querier adds a [configurable number of concurrent workers](https://github.com/cortexproject/cortex/blob/1797adfed2979f6096c3305b0dc9162c1ec0c046/pkg/querier/worker/worker.go#L212)
+which are each capable of executing a query.
+Therefore, scaling the query frontend impacts the amount of work each individual querier is attempting to do at any given time.
```
```suggestion
Each worker is connected to a single query-frontend instance, therefore scaling up the query-frontend impacts the amount of work each individual querier is attempting to do at any given time.
```

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,25 +5,52 @@ weight: 5
+`-querier.frontend-address` and `-querier.scheduler-address` options are mutually exclusive, and at most one can be set.
+
+When using query scheduler, it is recommended to run two query scheduler instances.
+Running only one query scheduler poses a risk of increased query latency when single scheduler crashes or restarts.
```
> poses a risk of increased query latency

Increased query latency or failing queries? If the query-frontend can't enqueue a query in the query-scheduler, doesn't the query fail?

pstibrany replied:
If frontend has no workers attempting to forward messages to scheduler (eg. because no scheduler is running), or frontend workers are unable to connect (send "INIT" message and get "OK" back), then frontend will get stuck while trying to pass query to the channel -- but no worker is reading from the channel. Once workers connect to schedulers, they will process the query. From user's perspective, query is just slower.

If workers are connected and read from channel, then forwarding of the query to scheduler may fail. In this case frontend will retry to send query to the channel. This adds latency. If this forwarding fails too many times, query fails.

pracucci commented on docs/operations/scalable-query-frontend.md:
```diff
@@ -5,25 +5,52 @@ weight: 5
+
+When using query scheduler, it is recommended to run two query scheduler instances.
+Running only one query scheduler poses a risk of increased query latency when single scheduler crashes or restarts.
+It shouldn't be necessary to run more than two schedulers.
```
```suggestion
Running two query-schedulers should be enough even for large Cortex clusters with an high QPS.
```

pracucci reviewed (approved):
Good job! I left very final non-blocking comments, but feel free to merge it without any other review from my side. Thanks!
